File: .agents/skills/search-israel-places/scripts/search_places.py
#!/usr/bin/env python3
"""OpenStreetMap/Overpass place search for Israel travel planning.

Usage:
    python search_places.py "<area>" --type restaurant|hotel|attraction [--max N]

Exits 0 and prints JSON array of place objects on success.
Exits 1 and prints {"error": "..."} on failure.

No API key required — uses free Nominatim + Overpass APIs.
"""

from __future__ import annotations

import json
import sys
import time
import urllib.parse
import urllib.request
import urllib.error

DEFAULT_MAX = 5
NOMINATIM_URL = "https://nominatim.openstreetmap.org/search"
OVERPASS_URL = "https://overpass-api.de/api/interpreter"
TIMEOUT = 25

# OSM tags for each category
TYPE_TAGS = {
    "restaurant": [
        '["amenity"="restaurant"]',
        '["amenity"="cafe"]',
    ],
    "hotel": [
        '["tourism"="hotel"]',
        '["tourism"="hostel"]',
        '["tourism"="guest_house"]',
        '["tourism"="apartment"]',
    ],
    "attraction": [
        '["tourism"="attraction"]',
        '["tourism"="museum"]',
        '["tourism"="viewpoint"]',
        '["tourism"="archaeological_site"]',
        '["historic"="ruins"]',
        '["historic"="archaeological_site"]',
        '["leisure"="nature_reserve"]',
    ],
}

TYPE_EMOJI = {
    "restaurant": "🍽️",
    "hotel": "🏨",
    "attraction": "📍",
}


def get_json(url: str, headers: dict | None = None) -> object:
    req = urllib.request.Request(url, headers=headers or {})
    req.add_header("User-Agent", "TrailMate/1.0 (travel planning agent)")
    with urllib.request.urlopen(req, timeout=TIMEOUT) as r:
        return json.loads(r.read().decode())


def post_json(url: str, body: str) -> object:
    req = urllib.request.Request(url, data=body.encode(), method="POST")
    req.add_header("User-Agent", "TrailMate/1.0 (travel planning agent)")
    with urllib.request.urlopen(req, timeout=TIMEOUT) as r:
        return json.loads(r.read().decode())


# ── Step 1: Geocode the area name ─────────────────────────────────────────────

def geocode(area: str) -> tuple[float, float, float, float]:
    """Return (south, west, north, east) bounding box for the area."""
    params = urllib.parse.urlencode({
        "q": area + ", Israel",
        "format": "json",
        "limit": 1,
        "countrycodes": "il",
    })
    results = get_json(f"{NOMINATIM_URL}?{params}")
    if not results:
        raise ValueError(f"Could not find area: {area}")
    r = results[0]
    lat, lon = float(r["lat"]), float(r["lon"])

    # Only trust the bbox when Nominatim found a proper polygon (way/relation).
    # For nodes (points) the bbox is fabricated and often wildly wrong.
    if r.get("osm_type") in ("way", "relation"):
        bbox = r.get("boundingbox", [])
        if len(bbox) == 4:
            s, n, w, e = float(bbox[0]), float(bbox[1]), float(bbox[2]), float(bbox[3])
            # Cap to at most ~50 km in each direction to avoid giant regions
            max_delta = 0.45
            if (n - s) <= max_delta * 2 and (e - w) <= max_delta * 2:
                return s, w, n, e

    # For nodes or oversized polygons: use center + sensible radius.
    # ~25 km radius for region names, ~10 km for city names.
    place_rank = r.get("place_rank", 20)
    delta = 0.23 if place_rank <= 12 else 0.10  # region vs city
    return lat - delta, lon - delta, lat + delta, lon + delta


# ── Step 2: Overpass search ───────────────────────────────────────────────────

def overpass_search(bbox: tuple, tags: list[str], max_results: int) -> list[dict]:
    s, w, n, e = bbox
    bbox_str = f"{s},{w},{n},{e}"

    # Build a union of all tag filters for nodes and ways
    parts = []
    for tag in tags:
        parts.append(f'node{tag}({bbox_str});')
        parts.append(f'way{tag}({bbox_str});')

    query = f"[out:json][timeout:{TIMEOUT}];({' '.join(parts)});out center tags {max_results * 3};"
    resp = post_json(OVERPASS_URL, query)
    return resp.get("elements", [])


# ── Step 3: Format results ────────────────────────────────────────────────────

def format_element(el: dict, place_type: str) -> dict | None:
    tags = el.get("tags", {})
    name = tags.get("name") or tags.get("name:en") or tags.get("name:he")
    if not name:
        return None  # skip unnamed places

    result: dict = {"name": name, "type": place_type}

    # Address
    addr_parts = []
    for key in ("addr:street", "addr:housenumber", "addr:city"):
        if v := tags.get(key):
            addr_parts.append(v)
    if addr_parts:
        result["address"] = ", ".join(addr_parts)
    elif city := tags.get("addr:city"):
        result["address"] = city

    # Location
    if "center" in el:
        result["location"] = {"lat": el["center"]["lat"], "lng": el["center"]["lon"]}
    elif el.get("type") == "node":
        result["location"] = {"lat": el["lat"], "lng": el["lon"]}

    # Contact
    if phone := tags.get("phone") or tags.get("contact:phone"):
        result["phone"] = phone
    if website := tags.get("website") or tags.get("contact:website"):
        result["website"] = website

    # Opening hours
    if hours := tags.get("opening_hours"):
        result["opening_hours"] = hours

    # Cuisine (restaurants)
    if place_type == "restaurant" and (cuisine := tags.get("cuisine")):
        result["cuisine"] = cuisine.replace(";", ", ")

    # Description / tourism info
    if desc := tags.get("description") or tags.get("description:en") or tags.get("wikipedia"):
        result["description"] = desc[:200]

    # OSM link
    osm_type = el.get("type", "node")
    osm_id = el.get("id")
    if osm_id:
        result["osm_url"] = f"https://www.openstreetmap.org/{osm_type}/{osm_id}"

    return result


def search_places(area: str, place_type: str, max_results: int) -> list[dict]:
    tags = TYPE_TAGS.get(place_type)
    if not tags:
        raise ValueError(f"Unknown type: {place_type}. Use restaurant, hotel, or attraction.")

    bbox = geocode(area)
    time.sleep(1)  # Nominatim rate limit
    elements = overpass_search(bbox, tags, max_results)

    results = []
    seen = set()
    for el in elements:
        place = format_element(el, place_type)
        if place and place["name"] not in seen:
            seen.add(place["name"])
            results.append(place)
        if len(results) >= max_results:
            break

    return results


# ── CLI ───────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    args = sys.argv[1:]
    if not args:
        print(json.dumps({"error": "Usage: search_places.py <area> --type restaurant|hotel|attraction [--max N]"}))
        sys.exit(1)

    area = args[0]
    place_type = "restaurant"
    max_results = DEFAULT_MAX

    i = 1
    while i < len(args):
        if args[i] == "--type" and i + 1 < len(args):
            place_type = args[i + 1]
            i += 2
        elif args[i] == "--max" and i + 1 < len(args):
            max_results = int(args[i + 1])
            i += 2
        else:
            i += 1

    try:
        results = search_places(area, place_type, max_results)
        print(json.dumps(results, ensure_ascii=False, indent=2))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)

File: web/lib/tools/searchPlaces.ts
// searchPlaces — restaurants / hotels / attractions via Nominatim + Overpass.
// Ported from .agents/skills/search-israel-places/scripts/search_places.py.
// No API key required.

import type { ToolDef } from "./types";

const NOMINATIM_URL = "https://nominatim.openstreetmap.org/search";
const OVERPASS_URL = "https://overpass-api.de/api/interpreter";
const TIMEOUT_S = 25;
const DEFAULT_MAX = 5;

const TYPE_TAGS: Record<string, string[]> = {
  restaurant: ['["amenity"="restaurant"]', '["amenity"="cafe"]'],
  hotel: [
    '["tourism"="hotel"]',
    '["tourism"="hostel"]',
    '["tourism"="guest_house"]',
    '["tourism"="apartment"]',
  ],
  attraction: [
    '["tourism"="attraction"]',
    '["tourism"="museum"]',
    '["tourism"="viewpoint"]',
    '["tourism"="archaeological_site"]',
    '["historic"="ruins"]',
    '["historic"="archaeological_site"]',
    '["leisure"="nature_reserve"]',
  ],
};

const UA = { "User-Agent": "TrailMate/1.0 (travel planning agent)" };

async function getJson(url: string): Promise<any> {
  const res = await fetch(url, { headers: UA });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

async function postJson(url: string, body: string): Promise<any> {
  const res = await fetch(url, { method: "POST", headers: UA, body });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

type BBox = [number, number, number, number]; // south, west, north, east

async function geocode(area: string): Promise<BBox> {
  const params = new URLSearchParams({
    q: `${area}, Israel`,
    format: "json",
    limit: "1",
    countrycodes: "il",
  });
  const results = await getJson(`${NOMINATIM_URL}?${params}`);
  if (!Array.isArray(results) || results.length === 0) {
    throw new Error(`Could not find area: ${area}`);
  }
  const r = results[0];
  const lat = parseFloat(r.lat);
  const lon = parseFloat(r.lon);

  // Trust the bbox only for polygons (way/relation); fabricate one for nodes.
  if (r.osm_type === "way" || r.osm_type === "relation") {
    const bb = r.boundingbox ?? [];
    if (bb.length === 4) {
      const s = parseFloat(bb[0]);
      const n = parseFloat(bb[1]);
      const w = parseFloat(bb[2]);
      const e = parseFloat(bb[3]);
      const maxDelta = 0.45;
      if (n - s <= maxDelta * 2 && e - w <= maxDelta * 2) return [s, w, n, e];
    }
  }
  const placeRank = r.place_rank ?? 20;
  const delta = placeRank <= 12 ? 0.23 : 0.1;
  return [lat - delta, lon - delta, lat + delta, lon + delta];
}

async function overpassSearch(bbox: BBox, tags: string[], maxResults: number) {
  const [s, w, n, e] = bbox;
  const bboxStr = `${s},${w},${n},${e}`;
  const parts: string[] = [];
  for (const tag of tags) {
    parts.push(`node${tag}(${bboxStr});`);
    parts.push(`way${tag}(${bboxStr});`);
  }
  const query = `[out:json][timeout:${TIMEOUT_S}];(${parts.join(" ")});out center tags ${maxResults * 3};`;
  const resp = await postJson(OVERPASS_URL, query);
  return resp.elements ?? [];
}

function formatElement(el: any, placeType: string): Record<string, any> | null {
  const tags = el.tags ?? {};
  const name = tags.name || tags["name:en"] || tags["name:he"];
  if (!name) return null;

  const result: Record<string, any> = { name, type: placeType };

  const addrParts: string[] = [];
  for (const key of ["addr:street", "addr:housenumber", "addr:city"]) {
    if (tags[key]) addrParts.push(tags[key]);
  }
  if (addrParts.length) result.address = addrParts.join(", ");
  else if (tags["addr:city"]) result.address = tags["addr:city"];

  if (el.center) result.location = { lat: el.center.lat, lng: el.center.lon };
  else if (el.type === "node") result.location = { lat: el.lat, lng: el.lon };

  if (result.location) {
    result.maps = `https://www.google.com/maps?q=${result.location.lat},${result.location.lng}`;
  }

  const phone = tags.phone || tags["contact:phone"];
  if (phone) result.phone = phone;
  const website = tags.website || tags["contact:website"];
  if (website) result.website = website;
  if (tags.opening_hours) result.opening_hours = tags.opening_hours;
  if (placeType === "restaurant" && tags.cuisine)
    result.cuisine = String(tags.cuisine).replace(/;/g, ", ");
  const desc = tags.description || tags["description:en"] || tags.wikipedia;
  if (desc) result.description = String(desc).slice(0, 200);

  if (el.id) result.osm_url = `https://www.openstreetmap.org/${el.type ?? "node"}/${el.id}`;
  return result;
}

export const searchPlaces: ToolDef = {
  schema: {
    type: "function",
    function: {
      name: "search_places",
      description:
        "Find real restaurants, hotels, or attractions in an Israeli area via " +
        "OpenStreetMap. Use to fill the eat/sleep parts of an itinerary. Returns " +
        "names, addresses, and Google Maps links.",
      parameters: {
        type: "object",
        properties: {
          area: { type: "string", description: "Area or city in Israel, e.g. 'Tiberias'." },
          type: {
            type: "string",
            enum: ["restaurant", "hotel", "attraction"],
            description: "What to search for.",
          },
          max: { type: "integer", description: "Max results (default 5)." },
        },
        required: ["area", "type"],
      },
    },
  },
  async execute(args: Record<string, any>) {
    const area = String(args.area ?? "").trim();
    const placeType = String(args.type ?? "restaurant").trim();
    const maxResults = Math.max(1, Math.min(15, Number(args.max ?? DEFAULT_MAX)));
    const tags = TYPE_TAGS[placeType];
    if (!area) return { status: "error", message: "area is required" };
    if (!tags) return { status: "error", message: `Unknown type: ${placeType}` };

    try {
      const bbox = await geocode(area);
      const elements = await overpassSearch(bbox, tags, maxResults);
      const results: Record<string, any>[] = [];
      const seen = new Set<string>();
      for (const el of elements) {
        const place = formatElement(el, placeType);
        if (place && !seen.has(place.name)) {
          seen.add(place.name);
          results.push(place);
        }
        if (results.length >= maxResults) break;
      }
      return { status: "success", places: results };
    } catch (e: any) {
      return { status: "error", message: String(e?.message ?? e) };
    }
  },
};
